packages/sanity/src/core/store/_legacy/document/utils/sequentializeListenerEvents.ts

```diff
@@ -0,0 +1,125 @@
+/* eslint-disable no-console */
+import {concat, from, type Observable, of, switchMap, throwError, timer} from 'rxjs'
+import {mergeMap, scan} from 'rxjs/operators'
+
+import {type ListenerEventWithSnapshot} from '../getPairListener'
+import {type MutationEvent} from '../types'
+import {discardChainTo, linkedSort} from './eventChainUtils'
+
+type State = {
+  base: {revision: string | undefined} | undefined
+  events: ListenerEventWithSnapshot[] | undefined
+  buffer: MutationEvent[]
+  initialSnapshotRevision: string | undefined
+}
+
+export class OutOfSyncError extends Error {
+  constructor(message: string) {
+    super(message)
+    this.name = 'OutOfSyncError'
+  }
+}
+
+const DEFAULT_MAX_BUFFER_SIZE = 10
```

Comment:
may be worth mentioning the hard coded value in buffered document in a comment above